onnx/onnx.proto

```diff
@@ -417,4 +420,78 @@ message OperatorSetIdProto {
   // The version of the operator set being identified.
   // This field MUST be present in this version of the IR.
   optional int64 version = 2;
+}
+
+// Defines a function parameter.
+message ParameterDeclProto {
+  // The name of this parameter. Must be present.
+  optional string name = 1;
+
+  // The type/shape of this parameter.
+  optional TypeProto type = 2; 
+
+  // A human-readable doc string for this parameter. 
+  optional string doc_string = 3; 
+} 
+
+// Defines a function attribute declaration.
+message FuncAttrDeclProto {
+  // The attribute name of the function. Must be present.
+  optional string name = 1;
+
+  // The type of this attribute.
+  optional AttributeProto.AttributeType type = 2;
+} 
+
+// Defines a custom implementation of a function. 
+message FunctionImplProto { 
+  // The domain of this function implementation.
+  optional string function_impl_domain = 1; 
+
+  // The version of this implementation.
+  optional int64 function_impl_version = 2; 
+
+  // The URI of the executable. A runtime uses <domain> and <version> to
+  // determine whether Compilation data itself saved in a separate file. 
+  optional string executable_uri = 3;
+
+  // The actual executable for this function implementation.
+  optional bytes executable = 4;
+} 
+ 
+// Defines a function.
+// We specify a function as a subgraph composed of ONNX ops. Any runtime can
+// have a custom implementation optimized for the runtime. A runtime can use
+// a matching custom implementation if there is one, or falls back to execute
+// the subgraph.
+// NOTE: We don't support recursive function.
+message FunctionDefProto { 
+  // The name of the function. MUST be present for this version of the IR. 
+  optional string name = 1;
+  
+  // The domain of the function.
+  optional string domain = 2;
+
+  // The version of the function.
+  optional int64 version = 3;
+
+  // The input parameters of the function. 
+  repeated ParameterDeclProto input_params = 4; 
+
+  // The output parameters of the function. 
+  repeated ParameterDeclProto output_params = 5;
+
+  // The body of the function. 
+  repeated NodeProto node = 6;
+
+  // The function attributes' declaration.
+  repeated FuncAttrDeclProto attr = 7;
```

Comment:
Are we removing attribute-references inside functions? If so, shouldn't we remove this field also? It can't really be used, right?